rfc/20250711-Middleware-For-Enhanced-Operations.md

```diff
@@ -0,0 +1,378 @@
+To provide users with a flexible and powerful way to extend OpenTofu's functionality, I propose implementing a middleware system that communicates with external processes via JSON-RPC 2.0 over STDIO. The middleware will be invoked at specific hook points during OpenTofu operations, allowing users to inspect, validate, and augment the behavior of resource operations without requiring changes to OpenTofu's core.
+
+### Hook Points for Operations
+The middleware system will provide hook points at two levels:
+
+**Resource-level hooks** (called for each individual resource/data source):
+- **pre-plan**
+- **post-plan**
+- **pre-apply**
+- **post-apply**
+- **pre-refresh**
+- **post-refresh**
+
+**Operation-level hooks** (called once per operation):
+- **init-stage-start**
+- **init-stage-complete**
+- **plan-stage-start**
+- **plan-stage-complete**
+- **apply-stage-start**
+- **apply-stage-complete**
+
+Additional hooks can be added in future versions as use cases emerge. It is possible that we should also consider hooks that handle failures/errors too.
+
+### Middleware Attachments
+
+For now I propose that middleware should be "attachable" to the entire project, or to a specific provider instance. This opens up the prospect of having provider-extension middleware or project-wide enforcement of rules.
+
+### Configuration Design
+Users will configure middleware through HCL blocks in their OpenTofu configuration:
+
+```hcl
+
+middleware "naming_convention_checker" {
+  command = "lint-opentofu-names"
+}
+
+middleware "cost_estimator" {
+  command = "python3"
+  args    = ["cost_estimator.py"]
+  env = {
+    API_KEY = var.cost_api_key
+  }
+}
+
+terraform { 
+  middleware = [middleware.naming_convention_checker] # Attach it across the entire project
+}
+
+provider "aws" {
+  middleware = [middleware.cost_estimator] # Or just attach it to a single provider
+  region = "us-east-1"
+}
+```
+
+### Middleware Execution Order
+
+When multiple middleware are attached to a provider or project, they execute in the order defined in the middleware array. This deterministic ordering allows for:
+
+- **Pipeline Processing**: Earlier middleware can validate/transform data for later middleware
+- **Priority Control**: Critical security checks can run before cost estimation
+- **Dependency Management**: Middleware that depends on others' metadata can run later
+
+Example execution flow:
+```hcl
+provider "aws" {
+  middleware = [
+    middleware.security_scanner,    # Runs first - can fail fast on violations
+    middleware.compliance_checker,  # Runs second - sees security metadata
+    middleware.cost_estimator      # Runs last - only if security/compliance pass
+  ]
+}
+```
+
+For pre-hooks (pre-plan, pre-apply, etc.), if any middleware returns a "fail" status, execution stops immediately and subsequent middleware in the chain are not called. For post-hooks, all middleware execute regardless of individual failures, allowing comprehensive metadata collection.
```

Comment:
I think we should allow the middleware to tell tofu what to do in both situations. I cant think of any use-cases off the top of my head, but maybe if some post hook wants to stop the execution for some valid reason we should allow them to do so?

Reply:
I think the hard part to handle is chaining middleware.  I assume that the reason for this is that you don't want chained middleware not finding data it needs in the input?